apps/testing-javascript/src/data/purchase-import.ts

```diff
@@ -0,0 +1,260 @@
+    ? NonNullable<BaseType[Key]>
+    : BaseType[Key]
+}
+
+type CleanMerchantCharge = SetNonNullable<MerchantCharge, 'merchantCustomerId'>
+
+const PurchaseSchema = z
+  .object({
+    id: z.string(),
+    userId: z.string(),
+    createdAt: z.string(),
+    totalAmount: z.coerce.number(),
+    city: z.string().nullable(),
+    state: z.string().nullable(),
+    country: z.string().nullable(),
+    ip_address: z.string().nullable(),
+    status: z.union([
+      z.literal('Valid'),
+      z.literal('Refunded'),
+      z.literal('Banned'),
+    ]),
+    productId: z.string(),
+    couponId: z.string().nullable(),
+    merchantChargeId: z.string().nullable(),
+    upgradedFromId: z.string().nullable(),
+    bulkCouponId: z.string().nullable(),
+    redeemedBulkCouponId: z.string().nullable(),
+    merchantPurchaseId: z.string().nullable(),
+    // quantity: z.number().nullable(),
+  })
+  .transform(({createdAt, merchantPurchaseId, ...rest}) => {
+    return {
+      createdAt: new Date(createdAt),
+      merchantSessionId: merchantPurchaseId,
+      ...rest,
+    }
+  })
+
+const FileDataSchema = z.object({
+  users: z.array(UserSchema),
+  coupons: z.array(CouponSchema),
+  merchantCustomers: z.array(MerchantCustomerSchema),
+  merchantCharges: z.array(MerchantChargeSchema),
+  purchases: z.array(PurchaseSchema),
+})
+
+const importPurchaseData = async () => {
+  // Make sure this only gets run against the KCD Products database
+  if (process.env.DATABASE_URL !== 'mysql://root@localhost:3309/kcd-products') {
+    console.log('This is only meant to run against the kcd-products database.')
+    process.exit(1)
+  }
+
+  // ***************************************
+  // ** Read and Parse Full Purchase JSON **
+  // ***************************************
+  let {users, coupons, merchantCustomers, merchantCharges, purchases} =
+    FileDataSchema.parse(readJsonData(dataFilePath))
+
+  const shortRun = false
+
+  const chunkData = <T>(data: Array<T>): Array<T>[] => {
+    const chunks = chunk(data, 1000)
+
+    if (shortRun) {
+      return chunks.slice(0, 1)
+    } else {
+      return chunks
+    }
+  }
+
+  // These UUIDs have no significance and aren't related to existing data. They
+  // were generated to be used as a consistent ID value for the User and
+  // Customer records.
+  const nullUserId = 'EFAEA7AA-2D37-481B-883E-E2D0DAF5ACD4' as string
+  const nullCustomerId = 'kcd_50A6E8C7-0ECA-43C1-848B-91080EAD7B4D' as string
+
+  // Create a null customer if it doesn't already exist
+  await prisma.user.create({
+    data: {
+      id: nullUserId,
+      name: 'Null Customer',
+      email: 'nullcustomer@egghead.io',
+      emailVerified: null,
+    },
+  })
+  await prisma.merchantCustomer.create({
+    data: {
+      id: nullCustomerId,
+      userId: nullUserId,
+      merchantAccountId: 'kcd_ff532118-69fe-4263-85a5-50b7b03a4b1e',
+      identifier: 'cus_kcd_null_customer',
+      createdAt: now,
+      status: 0,
+    },
+  })
```

Comment:
This script creates and uses a Null Customer for the many `MerchantCharge` records that would otherwise not be associated with a Stripe customer because of how purchases used to be set up.